lib/absinthe/phase/document/execution/resolution.ex

```diff
@@ -180,7 +180,7 @@ defmodule Absinthe.Phase.Document.Execution.Resolution do
 
   # bp_field needs to have a concrete schema node, AKA no unions or interfaces
   defp do_resolve_field(res, source, path) do
-    res
+    call_middleware({Absinthe.Middleware.Telemetry, []}, res)
```

Comment:
Right, this is actually kind of tricky. The hard part is two fold:

1) do_resolve_field is called on all fields, so we're back to having middleware run on all fields instead of just on fields with "full" resolvers
2) If however we simply:
```
defmacro resolve(fun) do
  quote do
    middleware Telemetry
    middleware Absinthe.Resolution, unquote(fun)
  end
end
```
this gets us most of the way there, but we run into trouble the minute people start messing with the middleware stack. For example if they add an auth middleware to the front of all resolvers on the root query object (a very common pattern) then the telemetry won't actually pick up the runtime cost of auth.

I should also note that we can't try to add middleware dynamically in a document blueprint phase due to how abstract types work. If a type is abstract then it the concrete fields with middleware won't actually be fetched until the resolution phase when we can determine the type, so consequently it isn't possible to rework the middleware list prior to resolution.

HOWEVER we may still have some options. I think we can do this when the schema gets built. If we go https://github.com/absinthe-graphql/absinthe/blob/master/lib/absinthe/schema/notation.ex#L1665 and in that third clause look through the list of middleware to see if `Absinthe.Resolution` is one of them, then we could prepend Telemetry to the front. We could even actually append the complete one as well and thus avoid the added complexity within the Resolution middlware.

I probably should have suggested this earlier but I only just thought of it!

Reply:
I put the middleware in `__ensure_middleware__` but I didn't split it into 2 middleware since it's actually very handy to mark the start_time in the first and send it in as options to the second

Reply:
Yup that's totally fine. I thought of the perfect way to make this apply to ONLY fields that use `resolve` while also making it possible to opt out.

Change https://github.com/absinthe-graphql/absinthe/blob/master/lib/absinthe/schema/notation.ex#L528 to 
```
defmacro resolve(fun_ast) do
  quote do
    meta :absinthe_telemetry, true
    middleware Absinthe.Resolution, unquote(fun_ast)
  end
end
```

And then in the `__ensure_middleware__` you can

```
def __ensure_middleware__(middleware, field, object) do
  if Absinthe.Type.meta(field, :absinthe_telemetry) do
    # prepend telemetry
  else
    middleware
  end
end
```

What's nice about this is that
1) By default it'll only happen for `resolve` based fields
2) However anyone could opt in easily
2) Opting out is easy too, you just configure `:absinthe_telemetry false` via any of the schema modification mechanisms.

Reply:
I can do that, however the opt out would only be partial, just for that specific resolver function, which is only a small part of the overall telemetry events.

I think telemetry is intended to be very low impact, so if there is no handler attached it's pretty much just an ets lookup and a function call or two

Reply:
ok, updated!